sample-xml/app/src/main/java/co/nimblehq/sample/xml/di/modules/PreferencesModule.kt

```diff
@@ -0,0 +1,32 @@
+package co.nimblehq.sample.xml.di.modules
+
+import android.content.Context
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.preferencesDataStore
+import co.nimblehq.sample.xml.data.repository.AppPreferencesRepositoryImpl
+import co.nimblehq.sample.xml.domain.repository.AppPreferencesRepository
+import dagger.Module
+import dagger.Provides
+import dagger.hilt.InstallIn
+import dagger.hilt.android.qualifiers.ApplicationContext
+import dagger.hilt.components.SingletonComponent
+import javax.inject.Singleton
+
+private val Context.appPreferencesDataStore by preferencesDataStore("app_preferences")
+
+@Module
+@InstallIn(SingletonComponent::class)
+class PreferencesModule {
+
+    @Provides
+    @Singleton
+    fun provideAppPreferencesDataStore(
+        @ApplicationContext applicationContext: Context
+    ): DataStore<Preferences> = applicationContext.appPreferencesDataStore
+
+    @Provides
+    fun provideAppPreferencesRepository(
+        appPreferencesDataStore: DataStore<Preferences>
+    ): AppPreferencesRepository = AppPreferencesRepositoryImpl(appPreferencesDataStore)
+}
```

Comment:
We can simply bind `AppPreferencesRepositoryImpl` to `AppPreferencesRepository` since it's already using the `@Inject` annotation, which can give [better performance](https://blog.mindorks.com/dependency-injection-with-dagger2-advance/#:~:text=This%20provides%20better%20performance.).
```suggestion
    @Binds
    abstract fun bindAppPreferencesRepository(
        pppPreferencesRepositoryImpl: AppPreferencesRepositoryImpl
    ): AppPreferencesRepository
}
```

Reply:
@AVI5HEK Thanks for your suggestion. Better to have this in the part1 tho. 😭 This is part2 that we apply the same implementation from the Part1 #389 .

Reply:
@AVI5HEK Wait, I'm afraid that we could not @Binds it because there will be both `@Provide` and `@Binds` inside the same module. So it will throw error like this:

<img width="1169" alt="Screenshot 2566-03-01 at 3 38 48 PM" src="https://user-images.githubusercontent.com/12026942/222087262-4e92d777-b85f-45bf-9c43-de06a835f3b6.png">

Or did I miss something? 🙏🏻

Reply:
@kaungkhantsoe Sorry, I don't have a permission to see your PR from gulf. 🙏🏻

Reply:
@lydiasama I was going to say the same thing as @kaungkhantsoe  did. You can put the concrete provider implementation inside a `companion object` to fix this error.

Reply:
Something like this @lydiasama
```
abstract class StorageModule {
      companion object {
         @Provides
         fun provideSecuredLocalStorage(@ApplicationContext context: Context) =
             EncryptedSharedPreferences(context)
     }

     @Singleton
     @Binds
     abstract fun provideAppPersistence(appPersistenceImpl: AppPersistenceImpl): AppPersistence
 }
```

Reply:
@AVI5HEK Thanks for your details. I got it now. 👍🏻

Reply:
@AVI5HEK @kaungkhantsoe Resolved at 4eb0acf in Part3 🙏🏻